test/IRGen/emulated-tls.swift

```
@@ -0,0 +1,15 @@
+// RUN: %target-swift-frontend -Xcc -femulated-tls %s -S -import-objc-header %S/Inputs/tls.h | %FileCheck %s --check-prefix=EMUTLS --check-prefix=EMUTLS-%target-os
+// RUN: %target-swift-frontend -Xcc -fno-emulated-tls %s -S -import-objc-header %S/Inputs/tls.h | %FileCheck %s --check-prefix=NOEMUTLS
+
+_swift_stdlib_gettid()
+
+// EMUTLS: __emutls_v._swift_stdlib_gettid.tid
+// EMUTLS-linux-android: __emutls_get_address
+// EMUTLS-linux-gnu: __emutls_get_address
+// EMUTLS-macosx: __emutls_get_address
+// EMUTLS-openbsd: __emutls_get_address
+// EMUTLS-windows-msvc: __emutls_get_address
+// EMUTLS-wasi-NOT: __emutls_get_address
```

Hmm, why is this negated for WASI? Perhaps we should warn about the ignored setting?

---

I honestly didn't want to deal with WASI at all, but was forced to because the linux CI builds it and runs these tests for it too.

My understanding is that the initial `__emutls_v._swift_stdlib_gettid.tid` symbol is swapped in because the LLVM codegen pass runs on all platforms, but it doesn't have `__emutls_get_address` in compiler-rt, like the other platforms do, so it doesn't call that function. I could've simply disabled this test for WASI instead, but I figured it was worth checking the `__emutls_v._swift_stdlib_gettid.tid` symbol at least on all platforms.

This is just an artifact of the current CI and WASI support, I don't think it's worth addressing in the test.

---

Ah, okay, its just that the call is something else? That seems reasonable enough then. Perhaps an issue to track restoring this would be good.

---

> Ah, okay, its just that the call is something else?

I don't know: I just did what I had to do to get the test to pass, as the CI was forcing me to address WASI but I'm not that familiar with it, didn't know if it even had threads yet. Looking at the relevant LLVM source now, [the `LowerToTLSEmulatedModel()` method that inserts the `__emutls_get_address` call](https://github.com/llvm/llvm-project/blob/112793a90ec84c31ced9bdacad7ce204e2148d2c/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L10512) is not invoked for the WASM target, as it is [for several other architectures](https://github.com/llvm/llvm-project/blob/1729e6e742ba9f6f210550000ace4bec72530c2e/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L10344). It appears that [WASI threads is still in the standardization phase](https://github.com/WebAssembly/proposals?tab=readme-ov-file#phase-4---standardize-the-feature-wg), so the implementation may still be in flux.

@kateinoigakukun, could you let us know if this might be a useful flag for WASM one day and if this test is okay for now?

---

All WebAssembly targets currently have proper TLS support (assuming dynamic linking is only for Emscripten, not with WASI), so the lowering does not implement emulation at all and silently ignores the option today as you mentioned. The thread representation model in Wasm is still unstable but I don't expect we need the emulation in the future, so it's fine as is now.